### ISSUE TITLE
Support HAL Forms actions with no request body

### ISSUE THREAD

**Issue by sazzer** [badgateway/ketting#453]
Unless I'm missing something, there's no way to specify a HAL Forms `_template` entry that doesn't send a request body. For example, to indicate in a HAL document that you can perform some action by doing a POST to some URI without a request body.

I can fake this using `_links` but it doesn't really feel right - they are meant to define links to other related resources, not actions that can be performed on this one.

For example, a User resource might look like this:
```
{
  "name": "Test User",
  "email": "testuser@example.com",
  "_links": {
    "self": {
      "href": "/users/auth0%7C6240a09cc665610070aa5bfb"
    }
  },
  "_templates": {
    "edit": {
      "method": "PUT",
      "fields": [
        {
          "name": "name",
          "type": "text",
          "required": true
        }, {
          "name": "email",
          "type": "email",
          "required": true
        }
      ]
    },
    "/users/actions/change_password": {
      "target": "/users/auth0%7C6240a09cc665610070aa5bfb/change_password",
      "method": "POST"
    },
    "/users/actions/verify_email": {
      "target": "/users/auth0%7C6240a09cc665610070aa5bfb/verify_email",
      "method": "POST"
    }
  }
}
```

In this case, both the `/users/actions/change_password` and `/users/actions/verify_email` templates define actions that are performed on this resource, but they don't need a request body. Instead they are simply a POST to the given URI that triggers the action and returns the result.

Doing the above will cause the action to send an `application/json` request with a request body of `{}`. Explicitly setting the `contentType` property on the template to `null` has no effect either. 

Further, because of the TS definition of `Action.submit()` I'm forced to pass in *some* payload when calling this method, so it seems that actions without a body are just not supported right now.

**Comment by sazzer** [comment]
I should mention that I only noticed this because Prism returns an HTTP 415 if the request has a body but the OpenAPI spec doesn't support one. Most server frameworks likely won't care about this.

**Comment by mamund** [comment]
the HAL-FORMS specification (here: https://rwcbook.github.io/hal-forms/) accounts for the ability to send body-less unsafe requests (see: https://rwcbook.github.io/hal-forms/#_code_properties_code).

FWIW, the HAL-FORMS example you show doesn't comply w/ the linked spec since the HAL-FORMS spec calls for a `properties` collection, not a `fields` collection. might that be a typo for this example? or is there another spec for FORMS in HAL that is in use here?

**Comment by sazzer** [comment]
That's a typo. I was converting it from Siren, which uses the "fields"
term, because HAL Forms has inbuilt support for required and readonly.

**Comment by mamund** [comment]
no problem. there are a couple "flavors" of HAL-FORMS out there, so i was just ticking a box there.

the real convo is about the body-less POST. the spec was _meant_ to support that. I assume the spec is OK but implementations may not be fully-compliant.

**Comment by evert** [comment]
Given:

> If the contentType property is missing, is set to empty, or contains an unrecognized value, the client SHOULD act is if the contentType is set to "application/json".

And:

> If the array is missing or empty, the properties collection MUST be treated as an empty set of parameters 

This tells me that we need to send valid application/json, and an empty set makes the most sense as `{}`.

**Comment by sazzer** [comment]
Both from HAL Forms? Maybe the problem is in that instead of Ketting then 🙂

As I said earlier, it's only really a problem because Prism rejects the
requests as having an unexpected content type in the request body...

**Comment by evert** [comment]
Yeah those are both excerpts =)

You could try application/x-www-form-urlencoded instead, as an empty set there is a 0-byte string!

**Comment by sazzer** [comment]
When not specifying a content type at all, Prism gives:
<img width="977" alt="image" src="https://user-images.githubusercontent.com/66072/193471307-e9d349cc-0e03-410e-b7bc-3dc6cc39b096.png">

And the Firefox console shows:
<img width="1439" alt="image" src="https://user-images.githubusercontent.com/66072/193471328-e8dc1833-2f1e-43fa-9eb8-b4c90bdc5b3a.png">

**Comment by evert** [comment]
I don't really know what Prism is, so not sure if that's normal :P is there anything else I could potentially do here?

**Comment by sazzer** [comment]
Interestingly - and I'm surprised by this - specifying `"contentType": "application/x-www-form-urlencoded"` in the HAL Template and it *does* work. :)

I really expected that to fail on the exact same error, since the Prism error was "Invalid content type", and this is still specifying a content-type header even if there's now no payload.

**Comment by sazzer** [comment]
> I don't really know what Prism is, so not sure if that's normal :P is there anything else I could potentially do here?

Sorry! Prism is this: https://stoplight.io/open-source/prism It's just a quick way of getting a server to produce responses for requests without yet having written a real server :)

**Comment by mamund** [comment]
I suspect it's a rule in prism that blindly attempts to parse a JSON object body when the `content-type: application/json` header is passed during _some_ HTTP methods (e.g. POST/PUT/PATCH) -- even if no body is included.

**Comment by sazzer** [comment]
Maybe. Regardless though, it seems that Ketting is working correctly *to the HAL Forms spec* - just that HAL Forms itself doesn't account for this use case - and I've got a workaround to make Prism happy enough until I actually get a real server working with these responses. So unless you have some desire to allow Ketting to support bodyless action requests this can be closed :)

**Comment by evert** [comment]
Personally I think if a `Content-Type` is set to `application/json` and no body is set, that _should_ be treated as invalid.

**Comment by evert** [comment]
Sorry, for the confusion, it was in response to @mamund :

> I suspect it's a rule in prism that blindly attempts to parse a JSON object body when the content-type: application/json header is passed during some HTTP methods (e.g. POST/PUT/PATCH) -- even if no body is included.


I think it's pretty good to expect for a tool to attempt to parse the body if there was a `Content-Type: application/json`

**Comment by mamund** [comment]
my view on this is that the Prism server has a bug. would be interested in how other frameworks handle the same use case.

FWIW, the HAL-FORMS spec uses *SHOULD*, not *MUST*. that means client apps can choose to _not_ send a `content-type` header with a body-less unssafe request.  

it might help to update the spec (here: https://rwcbook.github.io/hal-forms/#_code_contenttype_code) to say that, 

> if there is no body sent with the request (for some set of HTTP methods?) then the client application *MAY* send no `content-type` header.

or some better wordsmitthing thereof.

the remainder is up to client and server implementors to sort out.

**Comment by mamund** [comment]
@vert:

> I think it's pretty good to expect for a tool to attempt to parse the body if there was a Content-Type: application/json

agreed. but it might not be a good idea to reject the request. for example, `application/x-forms-urlencoded` is an example of what i think is the proper response.

and, as @sazzer points out, 

> It was sending a body though

that the body was `{}` and Prism is rejecting the request w/ 415. so it seems that Prism is getting a body that matches the stated `content-type` and _still_ rejecting the request. that seems buggy.

**Comment by evert** [comment]
I'd strongly suggest coding 1 specific behavior for this, otherwise it will never be possible to create a generic client that works with generic servers without out-of-band configuration

**Comment by sazzer** [comment]
Prism is rejecting the request not because the request is invalid, but
because it's invalid *with regards to the OpenAPI spec*.

The problem at that end is that the OpenAPI spec says that this request
shouldn't have a body but the client was sending one.

**Comment by mamund** [comment]
@sazzer 

good point.
